Cannot select the fake type when using pointers

Hey there, I am having trouble applying the faker:"Something" tag to struct fields that are pointers. 

Here is the struct: 
```javascript
type UserInput struct {
	FirstName         *string                 `json:"first_name,omitempty" faker:"first_name_male"`
	LastName          *string                 `json:"last_name,omitempty" faker:"first_name_male"`
	FullName          *string                 `json:"full_name,omitempty" faker:"first_name_male"`
	Email             *string                 `json:"email,omitempty" faker:"email"`
}
```

Stack:

panic: reflect.Set: value of type string is not assignable to type *string
```javascript
goroutine 1 [running]:
reflect.Value.assignTo(0x15f1b80, 0xc0001f6a60, 0x98, 0x16e569f, 0xb, 0x15e0520, 0x0, 0xc0001f6a60, 0x0, 0x0)
        /usr/local/Cellar/go/1.11.1/libexec/src/reflect/value.go:2269 +0x44c
reflect.Value.Set(0x15e0520, 0xc000210640, 0x196, 0x15f1b80, 0xc0001f6a60, 0x98)
        /usr/local/Cellar/go/1.11.1/libexec/src/reflect/value.go:1403 +0xa7
bitbucket.org/teammemento/demo-api/vendor/github.com/bxcodec/faker.setDataWithTag(0xc0000b6640, 0xc000210640, 0x16, 0x15f253e, 0xf, 0x16, 0x196)
        /Users/zkynet/go/src/bitbucket.org/teammemento/demo-api/vendor/github.com/bxcodec/faker/faker.go:311 +0x282
bitbucket.org/teammemento/demo-api/vendor/github.com/bxcodec/faker.getValue(0x17a2c20, 0x16ba440, 0x16ba440, 0x5, 0xc0002184f0, 0xc0000f7d18, 0x100dc0f)
        /Users/zkynet/go/src/bitbucket.org/teammemento/demo-api/vendor/github.com/bxcodec/faker/faker.go:209 +0x168b
bitbucket.org/teammemento/demo-api/vendor/github.com/bxcodec/faker.FakeData(0x161af20, 0xc0002105a0, 0x1d02880, 0x15d3c60)
        /Users/zkynet/go/src/bitbucket.org/teammemento/demo-api/vendor/github.com/bxcodec/faker/faker.go:152 +0x1f5
```

Now, if I remove the faker:"first_name_male" .. everything works fine. 

It probably has something to do with this change:
https://github.com/bxcodec/faker/pull/17/commits/d5d9f35bc27255714794d682d2163a65cd7cd531


Peace and love <3

--- a/address.go
+++ b/address.go
@@ -41,11 +41,8 @@ func (i Address) Latitude(v reflect.Value) (interface{}, error) {
 	kind := v.Kind()
 	val := i.latitute()
 	if kind == reflect.Float32 {
-		v.Set(reflect.ValueOf(val))
 		return float32(val), nil
 	}
-	v.Set(reflect.ValueOf(float64(val)))
-
 	return float64(val), nil
 }
 
@@ -58,9 +55,7 @@ func (i Address) Longitude(v reflect.Value) (interface{}, error) {
 	kind := v.Kind()
 	val := i.longitude()
 	if kind == reflect.Float32 {
-		v.Set(reflect.ValueOf(val))
 		return float32(val), nil
 	}
-	v.Set(reflect.ValueOf(float64(val)))
 	return float64(val), nil
 }

--- a/datetime.go
+++ b/datetime.go
@@ -652,7 +652,8 @@ func (d DateTime) UnixTime(v reflect.Value) (interface{}, error) {
 	var val int64
 	if kind == reflect.Int64 {
 		val = d.unixtime()
-
+	} else {
+		val = 0
 	}
 
 	return val, nil

--- a/datetime.go
+++ b/datetime.go
@@ -652,10 +652,9 @@ func (d DateTime) UnixTime(v reflect.Value) (interface{}, error) {
 	var val int64
 	if kind == reflect.Int64 {
 		val = d.unixtime()
-	} else {
-		val = 0
+
 	}
-	v.SetInt(val)
+
 	return val, nil
 }
 

--- a/datetime_test.go
+++ b/datetime_test.go
@@ -2,7 +2,6 @@ package faker
 
 import (
 	"fmt"
-	"log"
 	"reflect"
 	"testing"
 	"time"
@@ -25,19 +24,6 @@ func TestUnixTimeValueValid(t *testing.T) {
 
 	if time.Now().Unix() <= ref.some {
 		t.Error("UnixTime should return time <= now")
-	}
-}
-func TestUnixTimeValueNotValid(t *testing.T) {
-	d := GetDateTimer()
-	var ref = struct {
-		some int
-	}{
-		some: 1212,
-	}
-	d.UnixTime(reflect.ValueOf(&ref.some).Elem())
-	log.Println(ref.some)
-	if ref.some != 0 {
-		t.Errorf("UnixTime should return 0, get : %v", ref.some)
 	}
 }
 

--- a/faker.go
+++ b/faker.go
@@ -64,6 +64,50 @@ const (
 	SKIP               = "-"
 )
 
+var defaultTag = map[string]string{
+	Email:              Email,
+	MacAddress:         MacAddress,
+	DomainName:         DomainName,
+	URL:                URL,
+	UserName:           UserName,
+	IPV4:               IPV4,
+	IPV6:               IPV6,
+	PASSWORD:           PASSWORD,
+	CreditCardType:     CreditCardType,
+	CreditCardNumber:   CreditCardNumber,
+	LATITUDE:           LATITUDE,
+	LONGITUDE:          LONGITUDE,
+	PhoneNumber:        PhoneNumber,
+	TollFreeNumber:     TollFreeNumber,
+	E164PhoneNumber:    E164PhoneNumber,
+	TitleMale:          TitleMale,
+	TitleFemale:        TitleFemale,
+	FirstName:          FirstName,
+	FirstNameMale:      FirstNameMale,
+	FirstNameFemale:    FirstNameFemale,
+	LastName:           LastName,
+	NAME:               NAME,
+	UnixTime:           UnixTime,
+	DATE:               DATE,
+	TIME:               Time,
+	MonthName:          MonthName,
+	YEAR:               Year,
+	DayOfWeek:          DayOfWeek,
+	DayOfMonthTag:      DayOfMonth,
+	TIMESTAMP:          TIMESTAMP,
+	CENTURY:            CENTURY,
+	TIMEZONE:           TIMEZONE,
+	TimePeriodTag:      TimePeriod,
+	WORD:               WORD,
+	SENTENCE:           SENTENCE,
+	PARAGRAPH:          PARAGRAPH,
+	Currency:           Currency,
+	Amount:             Amount,
+	AmountWithCurrency: AmountWithCurrency,
+	ID:                 ID,
+	HyphenatedID:       HyphenatedID,
+}
+
 // TaggedFunction ...
 type TaggedFunction func(v reflect.Value) (interface{}, error)
 
@@ -207,6 +251,7 @@ func getValue(t reflect.Type) (reflect.Value, error) {
 					continue
 				default:
 					err := setDataWithTag(v.Field(i).Addr(), tag)
+					// err := setDataWithTag(t.Field(i).Type, tag)
 					if err != nil {
 						return reflect.Value{}, err
 					}
@@ -291,8 +336,32 @@ func setDataWithTag(v reflect.Value, tag string) error {
 		return errors.New(ErrValueNotPtr)
 	}
 
+	if _, exist := mapperTag[tag]; !exist {
+		return errors.New(ErrTagNotSupported)
+	}
+
 	v = reflect.Indirect(v)
 	switch v.Kind() {
+	case reflect.Ptr:
+		if _, def := defaultTag[tag]; !def {
+			res, err := mapperTag[tag](v)
+			if err != nil {
+				return err
+			}
+			v.Set(reflect.ValueOf(res))
+			return nil
+		}
+
+		res, err := mapperTag[tag](v)
+		if err != nil {
+			return err
+		}
+		t := v.Type()
+		newv := reflect.New(t.Elem())
+		rval := reflect.ValueOf(res)
+		newv.Elem().Set(rval)
+		v.Set(newv)
+		return nil
 	case reflect.Float32, reflect.Float64:
 		return userDefinedFloat(v, tag)
 	case reflect.String:
@@ -301,15 +370,11 @@ func setDataWithTag(v reflect.Value, tag string) error {
 	case reflect.Int, reflect.Int32, reflect.Int64, reflect.Int8, reflect.Int16:
 		return userDefinedInt(v, tag)
 	default:
-		if _, exist := mapperTag[tag]; !exist {
-			return errors.New(ErrTagNotSupported)
-		}
 		res, err := mapperTag[tag](v)
 		if err != nil {
 			return err
 		}
 		v.Set(reflect.ValueOf(res))
-
 	}
 	return nil
 }

--- a/faker.go
+++ b/faker.go
@@ -351,12 +351,12 @@ func setDataWithTag(v reflect.Value, tag string) error {
 			return nil
 		}
 
-		res, err := mapperTag[tag](v)
+		t := v.Type()
+		newv := reflect.New(t.Elem())
+		res, err := mapperTag[tag](newv.Elem())
 		if err != nil {
 			return err
 		}
-		t := v.Type()
-		newv := reflect.New(t.Elem())
 		rval := reflect.ValueOf(res)
 		newv.Elem().Set(rval)
 		v.Set(newv)

--- a/faker.go
+++ b/faker.go
@@ -251,7 +251,6 @@ func getValue(t reflect.Type) (reflect.Value, error) {
 					continue
 				default:
 					err := setDataWithTag(v.Field(i).Addr(), tag)
-					// err := setDataWithTag(t.Field(i).Type, tag)
 					if err != nil {
 						return reflect.Value{}, err
 					}

--- a/faker_test.go
+++ b/faker_test.go
@@ -547,3 +547,38 @@ func TestTagAlreadyExists(t *testing.T) {
 		t.Error("Expected ErrTagAlreadyExists Error,  But Got: ", err)
 	}
 }
+
+func TestTagWithPointer(t *testing.T) {
+	type TestStruct struct {
+		FirstName  *string  `json:"first_name,omitempty" faker:"first_name_male"`
+		Email      *string  `json:"email,omitempty" faker:"email"`
+		Latitude   *float64 `faker:"lat"`
+		Latitude32 *float32 `faker:"lat"`
+		UnixTime   *int64   `faker:"unix_time"`
+	}
+
+	var sample TestStruct
+	err := FakeData(&sample)
+	if err != nil {
+		t.Error("Expected Not Error, But Got: ", err)
+	}
+
+	//Assert
+	if sample.FirstName == nil || *sample.FirstName == "" {
+		t.Error("Expected filled but got emtpy")
+	}
+	if sample.Email == nil || *sample.Email == "" {
+		t.Error("Expected filled but got emtpy")
+	}
+	if sample.Latitude == nil || *sample.Latitude == 0 {
+		t.Error("Expected filled but got emtpy")
+	}
+	if sample.Latitude32 == nil || *sample.Latitude32 == 0 {
+		t.Error("Expected filled but got emtpy")
+	}
+
+	if sample.UnixTime == nil || *sample.UnixTime == 0 {
+		t.Error("Expected filled but got emtpy")
+	}
+
+}

--- a/faker_test.go
+++ b/faker_test.go
@@ -549,13 +549,19 @@ func TestTagAlreadyExists(t *testing.T) {
 }
 
 func TestTagWithPointer(t *testing.T) {
+
 	type TestStruct struct {
 		FirstName  *string  `json:"first_name,omitempty" faker:"first_name_male"`
 		Email      *string  `json:"email,omitempty" faker:"email"`
 		Latitude   *float64 `faker:"lat"`
 		Latitude32 *float32 `faker:"lat"`
 		UnixTime   *int64   `faker:"unix_time"`
+		School     *School  `faker:"school"`
 	}
+	// With custom provider
+	AddProvider("school", func(v reflect.Value) (interface{}, error) {
+		return &School{Location: "Jakarta"}, nil
+	})
 
 	var sample TestStruct
 	err := FakeData(&sample)
@@ -581,4 +587,7 @@ func TestTagWithPointer(t *testing.T) {
 		t.Error("Expected filled but got emtpy")
 	}
 
+	if sample.School == nil || sample.School.Location == "" {
+		t.Error("Expected filled but got emtpy")
+	}
 }

